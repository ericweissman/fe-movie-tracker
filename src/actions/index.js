export const getMovies = (movies) => ({
  type: 'GET_MOVIES',
  movies
});

export const loginUser = (user) => ({
  type: 'LOGIN_USER',
  user
})

export const logoutUser = () => ({
  type: 'LOGOUT_USER'
})

export const populateFavorites = (favorites) => ({
  type: 'POPULATE_FAVORITES',
  favorites
})
