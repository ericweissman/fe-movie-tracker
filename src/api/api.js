export const fetchData = async (url) => {
  const response = await fetch(url);
  if(response.ok) {
    return response.json()
  } else {
    throw Error(`Error fetching, ${response.status}`)
  }
}

export const postData = async (urlSuffix, data) => {
  const url = 'http://localhost:3000/api/users' + urlSuffix;
  const response = await fetch(url, {
    method: 'POST',
    body: JSON.stringify(data),
    headers: {
      'Content-Type': 'application/json'
    }
  })
  if(response.ok) {
    let data = response.json();
    return data
  } else {
    throw Error(`Error fetching, ${response.status}`)
  }
}