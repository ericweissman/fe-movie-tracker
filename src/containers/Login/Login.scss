@font-face {
  font-family: 'Cereal';
  src: url(./AirbnbCereal-Black.ttf)
}

.login-form {
  color: $text;
  align-items: center;
  border-radius: 10px;
  display: flex;
  flex-direction: column;
  margin: 100px auto;
  width: 400px;
  background-color: $accent;
  box-shadow: 7px 7px 5px -5px rgba(0,0,0,0.31);



  h2 {
    font-family: 'Cereal'
  }

  p {
    font-family: 'Roboto';
    color: $primary;
    text-transform: uppercase;
  }

  input {
    border: 2px solid rgb(240, 240, 240);
    border-radius: 10px;
    height: 30px;
    margin-bottom: 15px;
    padding-left: 5px;
    width: 200px;
    outline: none;
  }

  button {
    border: none;
    outline: none;
    color: $text;
    border-radius: 10px;
    cursor: pointer;
    font-family: 'Roboto', sans-serif;
    font-size: 1rem;
    height: 30px;
    margin-bottom: 15px;
    font-weight: 600;
    width: 100px;
    transition-duration: .15s;


    &:hover {
      background-color: $text;
      color: $primary;
    }
  }

  .link {
    font-family: 'Roboto', sans-serif;
    margin-bottom: 15px;
  }
}

@media only screen and (max-width: 449px) {
  .login-form {
    width: 285px
  }
}