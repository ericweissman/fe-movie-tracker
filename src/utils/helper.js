import {fetchData} from '../api/api'

export const getFavorites = async (id) => {
  const url = `http://localhost:3000/api/users/${id}/favorites`
  let userFavorites = await fetchData(url)
  userFavorites = userFavorites.data.map(favorite => {
    return favorite.movie_id
  })

  return userFavorites
}


