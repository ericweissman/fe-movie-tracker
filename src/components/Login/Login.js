import React, {Component} from 'react'
import { connect } from 'react-redux'
import { loginUser, populateFavorites } from '../../actions';
import { postData } from '../../api/api'
import { Redirect, Link } from 'react-router-dom';
import { getFavorites } from '../../utils/helper'
import { MovieArea } from '../MovieArea/MovieArea';

export class Login extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
      status: ''
    }
  }
  
handleChange = (event) => {
  const { name, value } = event.target
  this.setState({
    [name]: value
  })
}


 handleFavorites = async (id) => {
  let favoriteMovieIDs = await getFavorites(id)
  this.props.populateFavorites(favoriteMovieIDs)
  }


handleSubmit = async (event) => {
  const { email, password } = this.state;
  event.preventDefault();
  try {
    let currentUser = await postData('', { email, password })
    currentUser = {
      id: currentUser.data.id,
      name: currentUser.data.name,
      favorites: [],
      status: currentUser.status
    }
    this.props.loginUser(currentUser)
    await this.handleFavorites(currentUser.id)
    this.setState({ status: currentUser.status })
  } catch {
    this.setState({ status: 'error' })
  }
}

  render() {
    const { email, password, status } = this.state
    return (
      <form onSubmit={this.handleSubmit}>
        <h2>Login</h2>
        <input onChange={this.handleChange} name='email' value={email} placeholder='email'></input>
        <input onChange={this.handleChange} name='password' type='password' value={password} placeholder='password'></input>
        <button>Login</button>
        {status === 'error' && <p>Email and password do not match</p> }
        {status === 'success' && <Redirect to='/' /> }
        <Link to='/signup'>Sign Up</Link>
        {/* <MovieArea /> */}
      </form>
    )
  }
}

export const mapDispatchToProps = (dispatch) => ({
  loginUser: (user) => dispatch(loginUser(user)),
  populateFavorites: (favorites) => dispatch(populateFavorites(favorites))
})

export default connect(null, mapDispatchToProps)(Login)