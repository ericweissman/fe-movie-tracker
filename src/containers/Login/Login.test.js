import { Login, mapDispatchToProps } from './Login'
import React from 'react'
import { shallow } from 'enzyme';
import { loginUser, populateFavorites } from '../../actions';
import * as api from '../../api/api';
import * as helpers from '../../utils/helper'

jest.mock('../../api/api')
const loginUserMock = jest.fn();
const mockFavoriteIDs = [2, 3]
const populateFavoritesMock = jest.fn().mockImplementation(() => Promise.all(
  mockFavoriteIDs
))


describe('Login', () => {
  let wrapper;
  it('should match the snapshot', () => {
    wrapper = shallow(<Login />)
    expect(wrapper).toMatchSnapshot();
  });
  
  describe('handleChange', () => {
    it('should set state when handleChange is called', () => {
      const mockEvent = {target: {name: 'email', value:'a@a.com'}}
      const mockEvent2 = {target: {name: 'password', value:'x'}}
      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.state('email')).toEqual('a@a.com')
      wrapper.instance().handleChange(mockEvent2);
      expect(wrapper.state('password')).toEqual('x')
    });
    
  })

  describe('handleSubmit', () => {
    let wrapper;
    const loginUserMock = jest.fn();
    beforeEach(() => {
      wrapper = shallow(<Login loginUser={loginUserMock}/>)
      const mockEvent = { preventDefault: jest.fn() }
      wrapper.find('form').simulate('submit', mockEvent)
    });

    it('should call postData when the form is submitted', () => {
      expect(api.postData).toHaveBeenCalled()
    });

    it('should call loginUser with the correct params', async () => {
      const loginUserMock = jest.fn()
      wrapper = shallow(<Login loginUser={loginUserMock} />)
      const mockUser = {name: 'hill', id: 2};
      const mockData = { data: mockUser, status: 'success'}
      const mockCurrentUser = {
        id: 2,
        name: 'hill',
        favorites: [],
        status: 'success'
      }
      api.postData = await jest.fn(() => mockData)
      const mockEvent = { preventDefault: jest.fn() }
      await wrapper.instance().handleSubmit(mockEvent)
      expect(wrapper.instance().props.loginUser).toHaveBeenCalledWith(mockCurrentUser)
    });
  })

  describe('handleFavorites', () => {
    it('should call get favorites with the correct params', async () => {
      helpers.getFavorites = jest.fn(() => mockFavoriteIDs)
      const mockId = 2;
      wrapper.setProps({populateFavorites: populateFavoritesMock})
      wrapper.instance().handleFavorites(mockId)
      await expect(helpers.getFavorites).toHaveBeenCalledWith(mockId)
    });

    it('should call populate favorites with the correct params', async () => {
      helpers.getFavorites = jest.fn(() => mockFavoriteIDs)
      const mockId = 2;
      wrapper.setProps({ populateFavorites: populateFavoritesMock });
      await wrapper.instance().handleFavorites(mockId);
      expect(populateFavoritesMock).toHaveBeenCalled()
      expect(populateFavoritesMock).toHaveBeenCalledWith(mockFavoriteIDs)
      
    })
  })

  describe('mapDispatchToProps', () => {
    it('should call dispatch when using a function from mapDispatchToProps', () => {
      const mockDispatch = jest.fn();
      const user = { name: 'Hill', email: 'hill@hotmail.com', password: 'hill' }
      const favorites = [2, 4]
      const actionToDispatch = loginUser(user)
      const actionToDispatch2 = populateFavorites(favorites)
      const mappedProps = mapDispatchToProps(mockDispatch);
      
      mappedProps.loginUser(user)  
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)

      mappedProps.populateFavorites(favorites)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch2)
    })
  });
 

})