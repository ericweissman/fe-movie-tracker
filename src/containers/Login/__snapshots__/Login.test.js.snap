// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match the snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <h2>
    Login
  </h2>
  <input
    name="email"
    onChange={[Function]}
    placeholder="email"
    value=""
  />
  <input
    name="password"
    onChange={[Function]}
    placeholder="password"
    type="password"
    value=""
  />
  <button>
    Login
  </button>
  <Link
    replace={false}
    to="/signup"
  >
    Sign Up
  </Link>
</form>
`;
